### PR TITLE
boards: arm: Add CONFIG_PERSISTENT_GECKO_RTCC to efm32pg_stk3402a

### PR DESCRIPTION
**Pull request by DRuffer-tmo**
Add us to the CODEOWNERS of the efm32pg_stk3402a.
Add a Kconfig for turning on the COUNTER_GECKO_RTCC. Add counter to the twister yaml supported devices. Change the rtcc0 prescaler from 1 to 32768.
Set multiple CONFIG settings to maintain the clock. Add PERSISTENT_GECKO_RTCC to support this feature. Add list of SiLabs devices where this feature works. Set the RTCC manually in the shell's date set command.

### CODE REVIEW

File: drivers/counter/Kconfig.gecko [hunk 1]
```diff
@@ -1,17 +1,31 @@
...
+	  CNTMODE field set to NORMAL mode running at a 30.5 microsecond rate
+	  overflowing in 36.4 hours for Silicon Labs Gecko chips.
+
+config PERSISTENT_GECKO_RTCC
+	bool "Silicon Labs Gecko Time (RTCC) persistence"
+	default n
+	depends on DT_HAS_SILABS_GECKO_RTCC_ENABLED && SOC_FAMILY_EXX32
+	select SOC_GECKO_RMU
+	select SOC_GECKO_CMU
+	select SOC_GECKO_RTCC
```
Comment (fkokosinski) [inline drivers/counter/Kconfig.gecko:25]:
```suggestion
```

File: drivers/counter/Kconfig.gecko [hunk 2]
```diff
@@ -1,17 +1,31 @@
...
+	help
+	  Enable the Gecko RTCC module to preserve the counter operations
+	  across EM4 resets. This mode will also set the Posix time.
```
Comment (fkokosinski) [inline drivers/counter/Kconfig.gecko:28]:
```suggestion
	  across EM4 resets. This mode will also set the POSIX time.
```

File: drivers/counter/counter_gecko_rtcc.c [hunk 3]
```diff
@@ -11,9 +12,12 @@
 #include <errno.h>
 #include <zephyr/kernel.h>
 #include <zephyr/device.h>
+#include <zephyr/posix/time.h>
 #include <soc.h>
 #include <em_cmu.h>
+#include <em_rmu.h>
 #include <em_rtcc.h>
+#include <zephyr/sys/timeutil.h>
```
Comment (fkokosinski) [inline drivers/counter/counter_gecko_rtcc.c:20]: Please ifdef these new includes with `CONFIG_PERSISTENT_GECKO_RTCC`.

File: drivers/counter/counter_gecko_rtcc.c [hunk 4]
```diff
@@ -219,12 +223,68 @@ static uint32_t counter_gecko_get_pending_int(const struct device *dev)
...
+	defined(CONFIG_SOC_SERIES_EFM32PG12B)	/*!< EFM32 Pearl Gecko (Rev C onwards) */ || \
+	defined(CONFIG_SOC_SERIES_EFM32JG1$)	/*!< EFM32 Jade Gecko (Rev C onwards) */ || \
+	defined(CONFIG_SOC_SERIES_EFM32JG12$)	/*!< EFM32 Jade Gecko (Rev C onwards) */ || \
+	defined(CONFIG_SOC_SERIES_EFM32GG11$)	/*!< EFM32 Giant Gecko GG11 */ || \
+	defined(CONFIG_SOC_SERIES_EFM32GG12$)	/*!< EFM32 Giant Gecko GG12 */ || \
+	defined(CONFIG_SOC_SERIES_EFM32TG11$)	/*!< EFM32 Tiny Gecko 11 */)
+#define LB_CLW0           (*((volatile uint32_t *)(LOCKBITS_BASE)+122))
+#define LB_CLW0_PINBOOTLOADER    (1 << 1)
+#endif /* EFM32, EZR32 SOC Series */
+#endif /* PERSISTENT_GECKO_RTCC */
```
Comment (fkokosinski) [inline drivers/counter/counter_gecko_rtcc.c:275]: IMO this belongs in the SoC specific startup code, not in the driver itself.

File: drivers/counter/counter_gecko_rtcc.c [hunk 5]
```diff
@@ -279,6 +339,19 @@ static int counter_gecko_init(const struct device *dev)
 	CMU_ClockSelectSet(cmuClock_LFA, cmuSelect_LFXO);
 #endif
 
+#ifdef LB_CLW0
+	/* Turn off the bootloader enable bit to prevent RTCC changes */
+	LB_CLW0 &= !LB_CLW0_PINBOOTLOADER;
+#endif /* LB_CLW0 */
```
Comment (fkokosinski) [inline drivers/counter/counter_gecko_rtcc.c:345]: Ditto

File: drivers/counter/counter_gecko_rtcc.c [hunk 6]
```diff
@@ -279,6 +339,19 @@ static int counter_gecko_init(const struct device *dev)
...
+
+#ifdef CONFIG_PERSISTENT_GECKO_RTCC
+	/* Turn on reset limited mode to preserve RTCC counts */
+	RMU_ResetControl(rmuResetWdog, rmuResetModeLimited);
+	RMU_ResetControl(rmuResetCoreLockup, rmuResetModeLimited);
+	RMU_ResetControl(rmuResetSys, rmuResetModeLimited);
+	RMU_ResetControl(rmuResetPin, rmuResetModeLimited);
+#endif /* CONFIG_PERSISTENT_GECKO_RTCC */
```
Comment (fkokosinski) [inline drivers/counter/counter_gecko_rtcc.c:353]: Again - please move it to SoC/board platform code.

File: boards/arm/efm32pg_stk3402a/Kconfig.defconfig [hunk 7]
```diff
@@ -16,4 +18,9 @@ config CMU_HFXO_FREQ
 config CMU_LFXO_FREQ
 	default 32768
 
+config COUNTER_GECKO_RTCC
+	bool
+	default y
+	depends on COUNTER
```
Comment (fkokosinski) [inline boards/arm/efm32pg_stk3402a/Kconfig.defconfig:24]: Why is this needed? `COUNTER_GECKO_RTCC` should be enabled by the `DT_HAS_*_ENABLED` macro.

**Review by fkokosinski — CHANGES_REQUESTED** [review]

**Comment by fkokosinski** [comment]
Also - thanks a lot for separating out these changes! It makes it a lot easier to review :)

File: CODEOWNERS [hunk 8]
```diff
@@ -114,7 +114,7 @@
 /boards/arm/cy8ckit_062_wifi_bt/          @ifyall @npal-cy
 /boards/arm/cy8cproto_062_4343w/          @ifyall @npal-cy
 /boards/arm/disco_l475_iot1/              @erwango
-/boards/arm/efm32pg_stk3401a/             @rdmeneze
```
[the remark this answers is not in the document]
Comment (DRuffer-tmo) [reply CODEOWNERS:117]: Ah, I see what you are saying now. My mistake!

File: boards/arm/efm32pg_stk3402a/Kconfig.defconfig (hunk 7 above)
Comment (DRuffer-tmo) [reply boards/arm/efm32pg_stk3402a/Kconfig.defconfig:24]: I'm sorry, but I am not aware of the DT_HAS_*_ENABLED macro. Can you give me a link to where it is defined?

File: boards/arm/efm32pg_stk3402a/efm32pg_stk3402a_defconfig [hunk 9]
```diff
@@ -7,5 +9,10 @@ CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
 CONFIG_GPIO=y
+CONFIG_COUNTER=y
+CONFIG_POSIX_CLOCK=y
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=40000000
 CONFIG_CMU_HFCLK_HFXO=y
+CONFIG_PERSISTENT_GECKO_RTCC=y
+CONFIG_NOTIFY=y
+CONFIG_REBOOT=y
```
[the remark this answers is not in the document]
Comment (DRuffer-tmo) [reply boards/arm/efm32pg_stk3402a/efm32pg_stk3402a_defconfig:18]: At the moment, I don't remember why I needed CONFIG_NOTIFY, so I'll just remove it for now, but CONFIG_REBOOT provides a reliable way to test the RTCC persistence in the shell module, so I will put that into a new samples/subsys/shell/shell_module/boards/ folder. Is that ok?

File: boards/arm/efm32pg_stk3402a/Kconfig.defconfig (hunk 7 above)
Comment (DRuffer-tmo) [reply boards/arm/efm32pg_stk3402a/Kconfig.defconfig:24]: I see it's usage below and will see if I can fix this properly

File: drivers/counter/Kconfig.gecko [hunk 10]
```diff
@@ -1,17 +1,31 @@
...
 	help
-	  Enable counter driver based on RTCC module for Silicon Labs Gecko
-	  chips.
+	  Enable counter driver based on RTCC module with the RTCC_CTRL
+	  CNTMODE field set to NORMAL mode running at a 30.5 microsecond rate
+	  overflowing in 36.4 hours for Silicon Labs Gecko chips.
+
+config PERSISTENT_GECKO_RTCC
+	bool "Silicon Labs Gecko Time (RTCC) persistence"
+	default n
```
[the remark this answers is not in the document]
Comment (DRuffer-tmo) [reply drivers/counter/Kconfig.gecko:21]: Ok

File: drivers/counter/Kconfig.gecko (hunk 2 above)
Comment (DRuffer-tmo) [reply drivers/counter/Kconfig.gecko:28]: Ok

File: drivers/counter/Kconfig.gecko (hunk 1 above)
Comment (DRuffer-tmo) [reply drivers/counter/Kconfig.gecko:25]: How else are you suggesting that I enable RMU settings?

File: drivers/counter/Kconfig.gecko [hunk 11]
```diff
@@ -1,17 +1,31 @@
...
 # Copyright (c) 2019, Piotr Mienkowski
+# Copyright (c) 2022 T-Mobile USA, Inc.
+#
 # SPDX-License-Identifier: Apache-2.0
 
 config COUNTER_GECKO_RTCC
 	bool "Silicon Labs Gecko Counter (RTCC) driver"
 	default y
-	depends on DT_HAS_SILABS_GECKO_RTCC_ENABLED
+	depends on DT_HAS_SILABS_GECKO_RTCC_ENABLED && SOC_FAMILY_EXX32
```
[the remark this answers is not in the document]
Comment (DRuffer-tmo) [reply drivers/counter/Kconfig.gecko:11]: In both cases, I assume?

File: drivers/counter/counter_gecko_rtcc.c (hunk 4 above)
Comment (DRuffer-tmo) [reply drivers/counter/counter_gecko_rtcc.c:275]: In zephyrproject/zephyr/soc/arm/silabs_exx32/common/soc.c?
